commons/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java

```diff
@@ -0,0 +1,124 @@
+package org.infinispan.commons.configuration.attributes;
+
+import org.infinispan.commons.CacheConfigurationException;
+
+/**
+ *
+ * AttributeDefinition. Defines the characteristics of a configuration attribute. It is used to construct an actual {@link Attribute} holder.
+ *
+ * An attribute definition has the following characteristics:
+ * <ul>
+ * <li>A name</li>
+ * <li>A default value or a value initializer</li>
+ * <li>A type, which needs to be specified if it cannot be inferred from the default value, i.e. when it is null</li>
+ * <li>Whether an attribute is immutable or not, i.e. whether its value is constant after initialization or it can be changed</li>
+ * <li>A validator which intercepts invalid values</li>
+ * </ul>
+ *
+ * @author Tristan Tarrant
+ * @since 7.2
+ */
+public final class AttributeDefinition<T> implements Cloneable {
+   private final String name;
+   private final T defaultValue;
+   private boolean immutable;
+   private AttributeInitializer<T> initializer;
+   private final AttributeValidator<T> validator;
+   private final Class<?> type;
+
+   AttributeDefinition(String name, T initialValue, Class<T> type, boolean immutable, AttributeValidator<T> validator, AttributeInitializer<T> initializer) {
+      this.name = name;
+      this.defaultValue = initialValue;
+      this.immutable = immutable;
+      this.initializer = initializer;
+      this.validator = validator;
+      this.type = type;
+   }
+
+   public String name() {
+      return name;
+   }
+
+   @SuppressWarnings("unchecked")
+   public Class<T> getType() {
+      return (Class<T>) type;
+   }
+
+   public T getDefaultValue() {
+      return initializer != null ? initializer().initialize() : defaultValue;
+   }
+
+   public boolean isImmutable() {
+      return immutable;
+   }
+
+   public AttributeInitializer<T> initializer() {
+      return initializer;
+   }
+
+   public void setInitializer(AttributeInitializer<T> initializer) {
```

Comment:
It seems this is never invoked?  If so we might as well remove it and make the variable final instead.